pymeasure/instruments/lecroy/lecroyT3DSO1204.py

```diff
@@ -0,0 +1,1151 @@
+            - "attenuation": probe attenuation (float)
+            - "bandwidth_limit": bandwidth limiting enabled (bool)
+            - "coupling": "ac 1M", "dc 1M", "ground" coupling (str)
+            - "offset": vertical offset (float)
+            - "skew_factor": channel-tochannel skew factor (float)
+            - "display": currently displayed (bool)
+            - "unit": "A" or "V" units (str)
+            - "volts_div": vertical divisions (float)
+            - "inverted": inverted (bool)
+            - "trigger_coupling": trigger coupling can be "dc" "ac" "highpass" "lowpass" (str)
+            - "trigger_level": trigger level (float)
+            - "trigger_level2": trigger lower level for SLEW or RUNT trigger (float)
+            - "trigger_slope": trigger slope can be "negative" "positive" "window" (str)
+        """
+
+        ch_setup = {
+            "channel": self.number,
+            "attenuation": self.probe_attenuation,
+            "bandwidth_limit": self.bwlimit,
+            "coupling": self.coupling,
+            "offset": self.offset,
+            "skew_factor": self.skew_factor,
+            "display": self.display,
+            "unit": self.unit,
+            "volts_div": self.scale,
+            "inverted": self.invert,
+            "trigger_coupling": self.trigger_coupling,
+            "trigger_level": self.trigger_level,
+            "trigger_level2": self.trigger_level2,
+            "trigger_slope": self.trigger_slope
+        }
+        return ch_setup
+
+
+class LeCroyT3DSO1204(Instrument):
+    """ Represents the LeCroy T3DSO1204 Oscilloscope interface for interacting with the instrument.
+
+    Refer to the LeCroy T3DSO1204 Oscilloscope Programmer's Guide for further details about
+    using the lower-level methods to interact directly with the scope.
+
+    Attributes:
+        SLEEP_SECONDS: sleep time between commands. If a command is received less than
+        SLEEP_SECONDS after the previous one, the code blocks until at least SLEEP_SECONDS have
+        passed.
+        Because the oscilloscope takes a non neglibile time to perform some operations, it might
+        be needed for the user to tweak the sleep time between commands.
+        The SLEEP_SECONDS is set to 0.5 seconds as default just to be on the safe side. Its
+        optimal value heavily depends on the actual commands and on the connection type,
+        so it is impossible to give a unique value to fit all cases.
+
+    .. code-block:: python
+
+        scope = LeCroyT3DSO1204(resource)
+        scope.autoscale()
+        ch1_data_array, ch1_preamble = scope.download_data(source="C1", points=2000)
+        # ...
+        scope.shutdown()
+    """
+
+    _BOOLS = {True: "ON", False: "OFF"}
+
+    SLEEP_SECONDS = 0.5
+
+    def __init__(self, adapter, **kwargs):
+        super().__init__(adapter, "LeCroy T3DSO1204 Oscilloscope", **kwargs)
+        if hasattr(self.adapter, "connection") and self.adapter.connection is not None:
```

Comment:
It should be sufficient to check, whether the connection is not None, as every Adapter instance has a connection.

Reply:
Change applied in commit ff305f8558e4761d67aeeff8dec5ab152a1814c7

Reply:
Just to confirm: the `self.adapter.connection is not None` is necessary because otherwise the protocol tests fail?

If the answer is yes, I feel that's a bit like `ProtocolAdapter` leaking out into the general implementation, such a check should not be necessary as instruments "normally" always have a `connection` after instantiating `Instrument`.
 @bmoneke should we do something about that? Put a generic mock `connection` in the `ProtocolAdapter` or somesuch?

Reply:
If you just set parameters of the connection or clear them, a mock connection would be fine. Some devices, however, expect a response (e.g. because they call some special functions of the VISAAdapter like get_stb or query_binary_values...). In these cases, a mock adapter won't work.

I agree, that instruments should not be adjusted to make the ProtocolAdapter work.

The problem is, that you are not supposed to use normally the adapter's connection directly, because the `connection` depends on the specific adapter type.
A generic mock connection seems good, because that would care for many use cases. Otherwise we had to add several methods (clear, close, timeout property, get_stb...) individually.
Or we could add timeout as a property to the `Adapter`, which will call its connection depending on the Adapter type (ProtocolAdapter stores it as a variable)

I made issue #742 to discuss possible solutions (not in here).